### PR TITLE
Add an event to parse a form added to the DOM on the fly for showon.

### PR DESCRIPTION
**Pull request by cheesegrits**
Pull Request for Issue # .

### Summary of Changes

The 'showon' feature for forms doesn't currently provide a setup method when adding a new form to the DOM on the fly, rather than on domready.  There's a way to add a repeated instance to an existing form, which uses some additional data attributes to add repeat numbering, but my use case involves adding new forms in the backend through AJAX. 

This change just renames setUpShowon as Joomla.setUpShowon, so it can be called through the global Joomla object..

### Testing Instructions

Pick any backup param that has showon, like global settings "send mail".  Apply this change.  The showon feature still works.  :)

### Documentation Changes Required

None

### CODE REVIEW

File: media/jui/js/cms-uncompressed.js [hunk 1]
```diff
@@ -186,7 +186,12 @@ if (typeof(Joomla) === 'undefined') {
 		$(document).ready(function() {
 			setUpShowon();
 
-			// Setup showon feature in the subform field
+            // Provide a way to parse an entire form that has been added to the DOM on the fly
```
Comment (Quy) [inline media/jui/js/cms-uncompressed.js]: Fix indentation. Use tabs.

**Comment by Fedik** [comment]
your fix is incorrect,
`subform` script do not have `subform-container-add` event,

to fix what you asked you should call `setUpShowon` from your ajax script, but first setUpShowon need to be under `Joomla.setUpShowon` scope

**Comment by cheesegrits** [comment]
OK, so how do we achieve that?

   -- hugh

**Comment by Fedik** [comment]
for Joomla! 3:
you can make a pull request to rename `setUpShowon` to `Joomla.setUpShowon`, 
so you can use it in your script:
```javascript
function ajaxCallback() {
 // do your stuf ...
 var blablaContainer = ...;
 // call setUpShowon
 Joomla.setUpShowon(blablaContainer);
}
```

for Joomla! 4 there will be a custom events, see #16016 , 
so you will be able to do in your script:
```javascript
function ajaxCallback() {
 // do your stuf ...
 var blablaContainer = ...;
 // dispatch custom event to inform about DOM updates
 Joomla.Event.dispatch(blablaContainer, 'joomla:updated');
}
```

**Comment by cheesegrits** [comment]
@Fedik - hopefully this is now the right way to do it?

**Comment by Fedik** [comment]
@cheesegrits yes, looks good now 😉

**Comment by cheesegrits** [comment]
@Fedik - cool.

BTW, where's a good place to bounce some stuff off you about the event dispatcher in 4.0, as it relates to bootstrap?  I've run across an immediate use for it, related to the same issue I'm addressing here, of adding entire forms to the DOM.  Similar problem, that the bootstrap-init.js only runs on domready.  I've patched it locally in 4.0 to add your joomla:updated event handling, but that raised a problem with Joomla.loadOptions() ...

I guess I can go ahead an do a PR for those changes, and maybe get your input there?

**Comment by Fedik** [comment]
@cheesegrits you can check #16016 as starting point, there introduced `Joomla.Event.dispatch` in `core.js`
and #18864 as an example how it can work

> I guess I can go ahead an do a PR for those changes, and maybe get your input there?

I do not know about a plans for `bootstrap-init.js` in Joomla! 4, but yes, any help are welcome :wink:

**Comment by cheesegrits** [comment]
@Fedik - can you give the green light for this one?  Then I can submit the bootstrap-init one.

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 2423a5cf29f39d02b23429bde11c9e6410f0a06c<br /><br />on review,
this just rename `setUpShowon` to `Joomla.setUpShowon` without logic changes, 
and this changes allow to extension developers to easily use `showon` feature with their AJAX code, or other dynamic code<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18843">issues.joomla.org/tracker/joomla-cms/18843</a>.</sub>

**Comment by Fedik** [comment]
@cheesegrits but if you need it only for Joomla! 4, then there it is fixed https://github.com/joomla/joomla-cms/pull/18864/commits/b10572db37b95357dd85cc5dfe803d21b335a18d (in another way)

**Comment by ghost** [comment]
@cheesegrits any Statement on Comment above by @Fedik?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18843">issues.joomla.org/tracker/joomla-cms/18843</a>.</sub>

**Comment by cheesegrits** [comment]
@franz-wohlkoenig oops, sorry, misread his comment and didn't think it needed a response. I need this functionality in 3.x and 4.x.  So yes, still need this merged. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18843">issues.joomla.org/tracker/joomla-cms/18843</a>.</sub>

**Comment by Sophist-UK** [comment]
I have tested this item ✅ successfully on https://github.com/joomla/joomla-cms/pull/18843/commits/2423a5cf29f39d02b23429bde11c9e6410f0a06c

**Comment by ghost** [comment]
@Fedik can you please retest?

File: media/jui/js/cms-uncompressed.js [hunk 2]
```diff
@@ -184,9 +186,9 @@ if (typeof(Joomla) === 'undefined') {
 		 * Initialize 'showon' feature
 		 */
 		$(document).ready(function() {
-			setUpShowon();
+			Joomla.setUpShowon();
 
-			// Setup showon feature in the subform field
+            // Setup showon feature in the subform field
```
Comment (Quy) [inline media/jui/js/cms-uncompressed.js:191]: Fix indentation. Please use tabs.

**Comment by Quy** [comment]
Do you have to apply the same changes to `cms.js`?

**Comment by Sophist-UK** [comment]
Presumably cms.js is the compressed version.

**Comment by ghost** [comment]
@Fedik can you please retest?

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 2423a5cf29f39d02b23429bde11c9e6410f0a06c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18843">issues.joomla.org/tracker/joomla-cms/18843</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by HLeithner** [comment]
are we sure that setUpShowon only lives in the closure and is private till this PR?

**Comment by Fedik** [comment]
@HLeithner I did not seen it anywhere else in the core :wink:

**Comment by HLeithner** [comment]
@Fedik what happens if you execute setUponshown multiple times on body? Wouldn't the animation be executed several times?

**Comment by Fedik** [comment]
hm yes, probably :smiley: 

But the idea is to call it once per _new_ `container`, example when part of the page was updated dynamically, 
that the case with "subform row".

Currently this was "isolated". 
After this patch `Joomla.setUpShowon()` can be called by extension developers, who want update the form dynamically (via ajax example), and be able to use ShowOn feature.

**Comment by HLeithner** [comment]
Programmers are lazy, they will simply use the form as container... is there a simple way to prevent the double attach?

**Comment by Fedik** [comment]
I can think about 2 way:
- set some flag to the container itself, that should be easy
- set some flag per field with "showon", that more safe but may need a bit more work

Can be as a data attribute `data-showon-attached="true"` or something similar

**Comment by HLeithner** [comment]
Would be good, ```jQuery('input').data('shown-attached', true);``` should be easy to implement.

Can you do this?

**Comment by Fedik** [comment]
me? yea, I guess :)

I just thought, is it worth to do it for j3.
Because for j4 Joomla.setUpShowon no need, there already used `joomla.updated` event.

**Comment by HLeithner** [comment]
Removing RTC and hope @Fedik find time to implement jQuery('shown-attached').

**Comment by ghost** [comment]
set label ´Updates Requested´ again.

**Comment by ghost** [comment]
reminder: 
> Removing RTC and hope @Fedik find time to implement jQuery('shown-attached').

**Comment by Fedik** [comment]
I have forked this, please have a look #32061

**Comment by richard67** [comment]
Closing in favour of #32061 . Thanks @cheesegrits for this PR here as a base for the fix.

**Pull request closed without merging**